### ISSUE TITLE
Add in support to export the yaml for missing rancher-compose/docker-compose fields

### ISSUE THREAD

**Issue by deniseschannon** [rancher/rancher#6281]
As docker-compose/rancher-compose has added supported fields, we are missing the support within Rancher. 

After Rancher-compose has added support, there are additional areas that need to be updated

1. Adding the field to the container API
**2. Adding the field to the exported docker-compose.yml of the stack- This ticket is to track being able to export these fields to docker-compose.yml when viewing the config of the stack** 
3. Adding the field to be read/imported by rancher-compose 
4. UI (Once Cattle is complete, we can track the UI in a separate issue)

Known Fields that are missing.

Field | Description
---|---
 - [ ] --add-host=[]                   | Add a custom host-to-IP mapping (host:ip)
 - [ ]   --blkio-weight             |     Block IO (relative weight), between 10 and 1000
 - [ ]   --blkio-weight-device=[]      |  Block IO weight (relative device weight)
 - [ ]  --cpu-shares            |        CPU shares (relative weight)
 - [ ]  --cgroup-parent        |         Optional parent cgroup for the container
 - [ ]  --cpu-period             |       Limit CPU CFS (Completely Fair Scheduler) period
 - [ ]  --cpu-quota              |       Limit CPU CFS (Completely Fair Scheduler) quota
 - [ ]  --cpuset-cpus            |       CPUs in which to allow execution (0-3, 0,1)
 - [ ]  --cpuset-mems            |       MEMs in which to allow execution (0-3, 0,1)
 - [ ]  --device-read-bps=[]     |       Limit read rate (bytes per second) from a device
 - [ ]  --device-read-iops=[]    |       Limit read rate (IO per second) from a device
 - [ ]  --device-write-bps=[]    |       Limit write rate (bytes per second) to a device
 - [ ]  --device-write-iops=[]    |      Limit write rate (IO per second) to a device
 - [ ]  --dns-opt=[]            |        Set DNS options
 - [ ]  --dns-search=[]      |           Set custom DNS search domains
 - [ ]  --group-add=[]       |           Add additional groups to join
 - [ ]  --ip                     |       Container IPv4 address (e.g. 172.30.100.104)
 - [ ]  --ip6                 |          Container IPv6 address (e.g. 2001:db8::33)
 - [ ]  --ipc                 |          IPC namespace to use
 - [ ]  --isolation         |            Container isolation level
 - [ ]  --kernel-memory      |           Kernel memory limit
 - [ ]  --memory-reservation   |         Memory soft limit
 - [ ]  --memory-swap             |      Swap limit equal to memory plus swap: '-1' to enable unlimited swap
 - [ ]  --memory-swappiness=-1      |    Tune container memory swappiness (0 to 100)
 - [ ]  --net-alias=[]          |        Add network-scoped alias for the container
 - [ ]  --oom-kill-disable    |          Disable OOM Killer
 - [ ]  --oom-score-adj       |          Tune host's OOM preferences (-1000 to 1000)
 - [ ]  --read-only              |       Mount the container's root filesystem as read only
 - [ ]  --security-opt=[]      |         Security Options
 - [ ]  --shm-size             |         Size of /dev/shm, default value is 64MB
 - [ ]  --stop-signal=SIGTERM    |       Signal to stop a container, SIGTERM by default
- [ ]  --sysctl value        |        Sysctl options (default map[])
 - [ ]  --tmpfs=[]            |          Mount a tmpfs directory
 - [ ]  --ulimit=[]         |            Ulimit options
 - [ ]  --uts                 |          UTS namespace to usea

**Comment by deniseschannon** [comment]
The original task has been split into 4 issues:

Cattle Changes: #4708
Rancher-compose: #6280
Exporting yml: #6281
UI: #6282

**Comment by alena1108** [comment]
Looks like this PR already supports yml export changes: 

https://github.com/rancher/cattle/pull/2047

**Comment by sangeethah** [comment]
Tested with v1.2.0-pre4-rc4.

Exported yaml is missing the following fields that is set for a service using launch_config option -  blkioDeviceOptions

- [ ] --device-read-bps=[]    Limit read rate (bytes per second) from a device
- [ ] --device-read-iops=[]    Limit read rate (IO per second) from a device
- [ ] --device-write-bps=[]    Limit write rate (bytes per second) to a device
- [ ] --device-write-iops=[]    Limit write rate (IO per second) to a device

**Comment by sangeethah** [comment]
@joshwget , sample yml file generated for the new fields:

```
version: '2'
services:
  test554094:
    dns_opt:
    - abc
    cpu_quota: 20000
    ulimits:
    - hard: 1024
      name: cpu
      soft: 1024
    ipc_mode: host
    groud_add:
    - root
    memswap_limit: -1
    cpuset: '0'
    cpuset_mems: '0'
    oom_score_adj: 100
    cpu_period: 10000
    shm_size: 1024
    image: ibuildthecloud/helloworld:latest
    cpu_shares: 400
    kernel_memory: 6000000
    sysctls:
      net.ipv4.ip_forward: '1'
    dns_search:
    - abc
    oom_kill_disable: true
    memory_swappiness: 100
    tmpfs:
      /tmp: rw
    uts: host
    memory_reservation: 5000000
    stop_signal: SIGTERM
    blkio_weight: 1000
```

**Comment by sangeethah** [comment]
2 fields yet to be tested are cgroup and cgroup-parent fields.

**Comment by StrongMonkey** [comment]
can verify we have all the fields that rancher-compose requires

**Comment by joshwget** [comment]
@StrongMonkey 
- `tmpfs` should be a list of strings and just contain the paths
- `ulimits` looks a little different in Docker Compose (https://docs.docker.com/compose/compose-file/#/ulimits)
- `memory_reservation` -> `mem_reservation`
- `memory_swappiness` -> `mem_swappiness`

**Comment by StrongMonkey** [comment]
@joshwget Changed. Can you take all the keys in tmpfs and make it an array. In that way we can support user to add options for `tmpfs` in agent.(engine api allow user to pass a map to tmpfs) If that is not easy to do that, i can change it in cattle to make it a separate field.

**Comment by joshwget** [comment]
@StrongMonkey This is how it's done in Docker Compose. They might add support for adding options later, but for not I'd rather not break compatibility.

**Comment by sangeethah** [comment]
Tested with rancher server version - v1.2.0-pre4-rc5.

Following fields that were missing are now present in exported yml file:

 --device-read-bps=[] Limit read rate (bytes per second) from a device
 --device-read-iops=[] Limit read rate (IO per second) from a device
 --device-write-bps=[] Limit write rate (bytes per second) to a device
 --device-write-iops=[] Limit write rate (IO per second) to a device
 --read-only
 --security-opt=[]
 --cgroup

```
    device_write_bps:
      /dev/null: 200
    device_write_iops:
      /dev/sda: 3000
    device_read_bps:
      /dev/null: 4000
    device_read_iops:
      /dev/sda: 2000
    cgroup: def
    security_opt:
    - label=user:USER
    - label=role:ROLE
    read_only: true
```

**Comment by sangeethah** [comment]
Following changes suggested by @joshwget , is also available now:

```  
   tmpfs:
    - /tmp:rw
    mem_reservation: 5000000
    mem_swappiness: 100
    ulimits:
      cpu:
        hard: 1024
        soft: 1024
```
